cloudsync/providers/file.py

```diff
@@ -0,0 +1,504 @@
+        return self._namespace
+
+    @namespace.setter
+    def namespace(self, path):
+        if self._namespace == path:
+            return
+        log.info("set namespace %s", path)
+        self._namespace = path
+
+        with self._api():
+            if not os.path.exists(self._namespace):
+                os.mkdir(self._namespace)
+
+            if self._observer:
+                self._observer.stop()
+
+            self._observer = watchdog_observer()
+            self._observer.schedule(Watcher(self), path, recursive=True)
+            self._observer.start()
+
+    @property
+    def namespace_id(self):
+        return self._fpath_to_oid(self._namespace)
+
+    @namespace_id.setter
+    def namespace_id(self, oid):
+        self.namespace = self._oid_to_fpath(oid)
+
+    def connect_impl(self, creds):
+        log.debug("connect mock prov creds : %s", creds)
+
+        if not creds:
+            raise ex.CloudTokenError()
+
+        self._api("connect_impl", creds)
+
+        if self.connection_id is None or self.connection_id == "invalid":
+            return os.urandom(16).hex()
+
+        return self.connection_id
+
+    def get_quota(self):
+        if not self.connected:
+            raise ex.CloudDisconnectedError()
+        _total, used, free = shutil.disk_usage(self._namespace)
+        return {
+            "used": used,
+            "limit": free,
+            "login": "n/a"
+        }
+
+    def _api(self, *args, **kwargs):
+        return self
+
+    def __enter__(self):
+        pass
+
+    def __exit__(self, ty, exception, tb):
+        if isinstance(exception, FileNotFoundError):
+            raise ex.CloudFileNotFoundError
+        if isinstance(exception, FileExistsError):
+            raise ex.CloudFileExistsError
+
+    @property  # type: ignore
+    def latest_cursor(self):
+        return self._latest_cursor
+
+    @property  # type: ignore
+    def current_cursor(self):
+        return self._cursor
+
+    @current_cursor.setter  # type: ignore
+    def current_cursor(self, val):
+        if val is None:
+            val = self.latest_cursor
+        if not isinstance(val, int) and val is not None:
+            raise ex.CloudCursorError(val)
+        if val > (self.latest_cursor + 1):
+            raise ex.CloudCursorError("%s is not a valid cursor" % val)
+
+        self._cursor = val
+
+    def _convert_watchdog_event(self, event):
+        otype = OType.DIRECTORY if event.is_directory else OType.FILE
+        fpath = event.src_path
+        oid = self._fpath_to_oid(fpath)
+        exists = True
+        prior_oid = None
+
+        if hasattr(event, "dest_path"):
+            prior_oid = oid
+            fpath = event.dest_path
+            oid = self._fpath_to_oid(fpath)
+
+        if type(event) in (watchdog_events.DirDeletedEvent, watchdog_events.FileDeletedEvent):
+            exists = False
+
+        if type(event) in (watchdog_events.DirModifiedEvent, watchdog_events.FileModifiedEvent) and exists:
```

Comment:
firstly, it seems like exists must be true at this point in the code, so it seems redundant to have `and exists` in the condition. Secondly, you want to return None on all file and folder modifications events?

Reply:
folder only